dm/relay/binlog_writer.go

```diff
@@ -61,6 +73,59 @@ func NewBinlogWriter(logger log.Logger, relayDir string) *BinlogWriter {
 	return &BinlogWriter{
 		logger:   logger,
 		relayDir: relayDir,
+		errCh:    make(chan error, 1),
+	}
+}
+
+// run starts the binlog writer.
+func (w *BinlogWriter) run() {
+	var (
+		buf       = &bytes.Buffer{}
+		errOccurs bool
+	)
+
+	// writeToFile writes buffer to file
+	writeToFile := func() {
+		if buf.Len() == 0 {
+			return
+		}
+
+		if w.file == nil {
+			select {
+			case w.errCh <- terror.ErrRelayWriterNotOpened.Delegate(errors.New("file not opened")):
+			default:
+			}
+			errOccurs = true
+			return
+		}
+		n, err := w.file.Write(buf.Bytes())
+		if err != nil {
+			select {
+			case w.errCh <- terror.ErrBinlogWriterWriteDataLen.Delegate(err, n):
+			default:
+			}
+			errOccurs = true
+			return
+		}
+		buf.Reset()
+	}
+
+	for bs := range w.input {
+		if errOccurs {
+			continue
+		}
+		if bs != nil {
+			buf.Write(bs)
```

Comment:
this will cause the byte content copied from `bs` to `buf`? not sure if use `[][]bytes` to take ownership of `bs` is faster

Reply:
0.43s/30s in frame graph, leave it for better code reading